x-pack/plugins/fleet/server/services/epm/packages/bulk_install_packages.ts

```diff
@@ -7,58 +7,71 @@
-import { upgradePackage } from './install';
+import { installPackage } from './install';
 import type { BulkInstallResponse, IBulkInstallPackageError } from './install';
 
 interface BulkInstallPackagesParams {
   savedObjectsClient: SavedObjectsClientContract;
-  packagesToUpgrade: string[];
+  packagesToInstall: string[];
   esClient: ElasticsearchClient;
 }
 
 export async function bulkInstallPackages({
   savedObjectsClient,
-  packagesToUpgrade,
+  packagesToInstall,
   esClient,
 }: BulkInstallPackagesParams): Promise<BulkInstallResponse[]> {
-  const installedAndLatestPromises = packagesToUpgrade.map((pkgToUpgrade) =>
-    Promise.all([
-      getInstallationObject({ savedObjectsClient, pkgName: pkgToUpgrade }),
-      Registry.fetchFindLatestPackage(pkgToUpgrade),
-    ])
+  const logger = appContextService.getLogger();
+  const installSource = 'registry';
+  const latestPackagesResults = await Promise.allSettled(
+    packagesToInstall.map((packageName) => Registry.fetchFindLatestPackage(packageName))
   );
-  const installedAndLatestResults = await Promise.allSettled(installedAndLatestPromises);
-  const installResponsePromises = installedAndLatestResults.map(async (result, index) => {
-    const pkgToUpgrade = packagesToUpgrade[index];
-    if (result.status === 'fulfilled') {
-      const [installedPkg, latestPkg] = result.value;
-      return upgradePackage({
-        savedObjectsClient,
-        esClient,
-        installedPkg,
-        latestPkg,
-        pkgToUpgrade,
-      });
-    } else {
-      return { name: pkgToUpgrade, error: result.reason };
-    }
-  });
-  const installResults = await Promise.allSettled(installResponsePromises);
-  const installResponses = installResults.map((result, index) => {
-    const pkgToUpgrade = packagesToUpgrade[index];
-    if (result.status === 'fulfilled') {
-      return result.value;
-    } else {
-      return { name: pkgToUpgrade, error: result.reason };
-    }
-  });
 
-  return installResponses;
+  logger.debug(`kicking off bulk install of ${packagesToInstall.join(', ')} from registry`);
+  const installResults = await Promise.allSettled(
+    latestPackagesResults.map(async (result, index) => {
+      const packageName = packagesToInstall[index];
+      if (result.status === 'fulfilled') {
+        const latestPackage = result.value;
+        return {
+          name: packageName,
+          version: latestPackage.version,
+          result: await installPackage({
+            savedObjectsClient,
+            esClient,
+            pkgkey: Registry.pkgToPkgKey(latestPackage),
+            installSource,
+            skipPostInstall: true,
+          }),
+        };
+      }
+      return { name: packageName, error: result.reason };
+    })
+  );
+
+  // only install index patterns if we installed any package for the first time
+  if (
+    installResults.find(
+      (result) => result.status === 'fulfilled' && result.value.result?.status === 'installed'
```

Comment:
Couple questions:

1. I haven't looked thoroughly at how installIndexPatterns works but if an upgrade of a package includes new mapping changes will those get applied if we don't call `installIndexPatterns` here? Would we not want to update the index patterns with the new mappings from the package?
2. Will the `status` field have values other than `installed`?

When I look at this code: https://github.com/elastic/kibana/blob/c91092fffdf954b05b7089efe219e9e7de24fbe0/x-pack/plugins/fleet/server/services/epm/packages/install.ts#L249

wouldn't the return value in a successful scenario of an upgrade or an install always be `installed`?

So I think this would return true if a package was installed for the first time or upgraded right?

Reply:
`status` can be `installed` or `already_installed`. When a package is successfully upgraded, the status will be `installed`, so this condition will be true, meaning that we will install index patterns. It only returns `already_installed` when the same package version is requested to be installed: https://github.com/elastic/kibana/blob/c91092fffdf954b05b7089efe219e9e7de24fbe0/x-pack/plugins/fleet/server/services/epm/packages/install.ts#L209

Reply:
Oh I see, sorry for the misunderstanding!